WebView/app/src/androidTest/java/com/android/samples/webviewdemo/MainActivityTest.kt

```diff
@@ -0,0 +1,158 @@
+        mainActivityRule.getActivity()
+        onWebView()
+            .withElement(findElement(Locator.ID, "location"))
+            .perform(webClick()) // Similar to perform(click())
+            .withElement(findElement(Locator.ID, "SF"))
+            .perform(webClick()) // Similar to perform(click())
+            .withElement(findElement(Locator.ID, "title"))
+            .check(webMatches(getText(), containsString("San Francisco")))
+    }
+
+    // Test for checking createJsObject
+    @Test
+    fun jsObjectIsInjectedAndContainsPostMessage() {
+    mainActivityRule.getActivity()
+    onWebView()
+        .check(
+            webMatches(
+                script("return jsObject && jsObject.postMessage != null;", castOrDie(Boolean::class.javaObjectType)),
+                `is`(true)
+            )
+        )
+    }
+
+    @Test
+    fun valueInCallback_compareValueInput_returnsTrue(){
+        mainActivityRule.getActivity()
+        // Setup
+        val jsObjName = "jsObject"
+        val allowedOriginRules = setOf<String>("https://example.com")
+        val message = "hello"
+        // Get a handler that can be used to post to the main thread
+        val mainHandler = Handler (Looper.getMainLooper());
+        val myRunnable = Runnable() {
+            val webView = WebView(context)
+            // Create JsObject
+            createJsObject(
+                webView,
+                jsObjName,
+                allowedOriginRules
+            ) { message -> //save message; call .set()
+            }
+            run() {
+                //Inject JsObject into Html
+                webView.loadDataWithBaseURL("https://example.com","<html></html>",
+                    "text/html", "UTF-8","https://example.com")
+                //Call js code to invoke callback
+                webView.evaluateJavascript("${jsObjName}.postMessage(${message})", null)
+            }
+        }
+        mainHandler.post(myRunnable)
+        // evaluate what comes out -> it should be hello
+        // *Note: //.get() is a place holder
+        assertEquals("hello"
+            , "//.get()"
+             )
+    }
+
+    @Test
+    // Checks that postMessage runs on the UI thread
+    fun checkingThreadCallbackRunsOn() {
+        mainActivityRule.getActivity()
+        // Setup
+        val jsObjName = "jsObject"
+        val allowedOriginRules = setOf<String>("https://example.com")
+        val message = "hello"
+        // Get a handler that can be used to post to the main thread
+        val mainHandler = Handler (Looper.getMainLooper())
+        // Start Interacting with webView on UI thread
+        val myRunnable = Runnable() {
+            run() {
+                val webView = WebView(context)
+                // Create JsObject
+                createJsObject(
+                    webView,
+                    jsObjName,
+                    allowedOriginRules
+                ) { message -> assertTrue(isUiThread()) }
```

Comment:
The next thing we'll want to do is to use a `Future<Boolean>` to store this value, and then call `assertTrue(future.get())` after we call Handler.post()